✨ Structured results for permissions

laurentsimon opened the pull request:
This PR is the first of a series to implement the structured results. It starts with the permission check.

This PR is very large because it includes not just the permission check's changes, but also all the needed changes to incorporate the structured results. Fortunately, the majority of files changes are small fixes to change package names.

The main changes to review:
1. `rule/` package. This takes a directory with a list of *.yml files. Each yml file contains the information for a rule. 
2. A check is composed of multiple rules
3. `finding` is a finding. It uses the `rule` to create a finding, and additional information such as a location, snippet, etc. provided by a check.This is eventually displayed to users
4. `pkg` is updated to use the structured results
5. A new format `extended-json` for the `--format` option is used. The `SCORECARD_EXPERIMENTAL=1` must be set to enable it.

An example result looks like the following:
`SCORECARD_EXPERIMENTAL=1 go run . --repo=GoogleCloudPlatform/rad-lab --format extended-json --show-details --checks Token-Permissions  | jq`
```json
{
  "date": "2023-01-06",
  "repo": {
    "name": "github.com/GoogleCloudPlatform/rad-lab",
    "commit": "4a94eb550a8fee9fd2e57a0ddaa70ded65903acb"
  },
  "scorecard": {
    "version": "(devel)",
    "commit": "unknown"
  },
  "score": 0,
  "checks": [
    {
      "risk": "High",
      "outcome": "Negative",
      "score": 0,
      "reason": "non read-only tokens detected in GitHub workflows",
      "name": "Token-Permissions",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions",
        "short": "Determines if the project's workflows follow the principle of least privilege."
      }
      "findings": [
        {
          "rule": "GitHubWorkflowPermissionsTopNoWrite",
          "outcome": "Negative",
          "risk": "High",
          "message": "topLevel 'contents' permission set to 'write'",
          "location": {
            "type": 1,
            "value": ".github/workflows/build-module-readme.yml",
            "lineStart": 28,
            "snippet": "write"
          },
          "remediation": {
            "text": "Update your workflow using https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions",
            "markdown": "Update your workflow using [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions).",
            "effort": "Low"
          }
        },
        {
          "rule": "GitHubWorkflowPermissionsTopNoWrite",
          "outcome": "Positive",
          "risk": "High",
          "message": "topLevel 'contents' permission set to 'read'",
          "location": {
            "type": 1,
            "value": ".github/workflows/check-tf-plan.yml",
            "lineStart": 30
          }
        },
...
    

```

There are several TODOS in the code which I will create a tracking issue for once the review is complete.

```release-note
Structured results for permissions
```

spencerschrock commented:
Still need to go through this, but WDUT about using the new extended-json format to implement #2577. Rules could also have numeric IDs?

A new results format would be a good time to add it since backwards compatibility isn't a concern.

laurentsimon commented:
> Still need to go through this, but WDUT about using the new extended-json format to implement #2577. Rules could also have numeric IDs?
> 
> A new results format would be a good time to add it since backwards compatibility isn't a concern.

The rule names are essentially ruleID that are more explicit than numbers. I can definitely add a number, but we have to be sure we don't get collision, ie we need some proper pre-submit to verify this over time.

joycebrum commented:
Regarding the 

```JSON
...
"remediation": {
  "text": "Update your workflow using https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions",
  "markdown": "Update your workflow using [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions).",
  "effort": "Low"
}
...
```
Is it possible to also suggest a direct remediation instruction? It has happened that a maintainer didn't like the tool and got confused about which options to enable to fix the check and though the remediation was also saying to enable Harden Runner which he didn't seem to want.

Maybe with a remediation like "Update your workflow by adding `contents: read` permission or using [...]", we would not rely fully on the step security tool and the maintainer can rather fix himself.

In this case, for example, the https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions does not fix the `contents: write` permission to `contents: read`, so the maintainer could ended up confused on what the check would be expecting from him since the stepsecurity itself didn't change anything.

![step security tool not changing the contents write option to read](https://user-images.githubusercontent.com/22223372/211325605-546a64a9-2586-475b-99ec-2cdf00431937.png)

In the print above I've selected the pin dependencies just to show that the changes should be applied.

laurentsimon commented:
> Regarding the
> 
> ```json
> ...
> "remediation": {
>   "text": "Update your workflow using https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions",
>   "markdown": "Update your workflow using [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions).",
>   "effort": "Low"
> }
> ...
> ```
> 
> Is it possible to also suggest a direct remediation instruction? It has happened that a maintainer didn't like the tool and got confused about which options to enable to fix the check and though the remediation was also saying to enable Harden Runner which he didn't seem to want.
> 
> Maybe with a remediation like "Update your workflow by adding `contents: read` permission or using [...]", we would not rely fully on the step security tool and the maintainer can rather fix himself.
> 
> In this case, for example, the https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions does not fix the `contents: write` permission to `contents: read`, so the maintainer could ended up confused on what the check would be expecting from him since the stepsecurity itself didn't change anything.
> 
> ![step security tool not changing the contents write option to read](https://user-images.githubusercontent.com/22223372/211325605-546a64a9-2586-475b-99ec-2cdf00431937.png)
> 
> In the print above I've selected the pin dependencies just to show that the changes should be applied.

Thanks for the feedback. One issue of asking users to fix the top-level themselves is that some of the steps may require write permissions, so their workflow would break after an update. If I update the remediation to explicitly ask to users to untick harden runner and pinned deps, would it be OK?

laurentsimon commented:
> > Regarding the
> > ```json
> > ...
> > "remediation": {
> >   "text": "Update your workflow using https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions",
> >   "markdown": "Update your workflow using [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions).",
> >   "effort": "Low"
> > }
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is it possible to also suggest a direct remediation instruction? It has happened that a maintainer didn't like the tool and got confused about which options to enable to fix the check and though the remediation was also saying to enable Harden Runner which he didn't seem to want.
> > Maybe with a remediation like "Update your workflow by adding `contents: read` permission or using [...]", we would not rely fully on the step security tool and the maintainer can rather fix himself.
> > In this case, for example, the https://app.stepsecurity.io/secureworkflow/GoogleCloudPlatform/rad-lab/build-module-readme.yml/main?enable=permissions does not fix the `contents: write` permission to `contents: read`, so the maintainer could ended up confused on what the check would be expecting from him since the stepsecurity itself didn't change anything.
> > ![step security tool not changing the contents write option to read](https://user-images.githubusercontent.com/22223372/211325605-546a64a9-2586-475b-99ec-2cdf00431937.png)
> > In the print above I've selected the pin dependencies just to show that the changes should be applied.
> 
> Thanks for the feedback. One issue of asking users to fix the top-level themselves is that some of the steps may require write permissions, so their workflow would break after an update. If I update the remediation to explicitly ask to users to untick harden runner and pinned deps, would it be OK?

@varunsh-coder there seems to be a bug in the webpage: the `contents: write` is not removed? Is it because you don't know which permissions the Actions used in steps require?

joycebrum commented:
> Thanks for the feedback. One issue of asking users to fix the top-level themselves is that some of the steps may require write permissions, so their workflow would break after an update. If I update the remediation to explicitly ask to users to untick harden runner and pinned deps, would it be OK?

I think it would still be fully dependent of the tool, but it would be much better because it will be clear to the user what he was supposed to use in the tool to fix the check. Something like 

```md
Update your workflow using the "Restrict permissions for GITHUB_TOKEN" at <link>
```  

Or maybe adding the "fix it yourself as a turn around" with something like: 

```md
Update your workflow using <link> or set the permission default to read (`contents: read`) giving write permissions only to jobs
```  

Both options I think would be great.

laurentsimon commented:
> > Thanks for the feedback. One issue of asking users to fix the top-level themselves is that some of the steps may require write permissions, so their workflow would break after an update. If I update the remediation to explicitly ask to users to untick harden runner and pinned deps, would it be OK?
> 
> I think it would still be fully dependent of the tool, but it would be much better because it will be clear to the user what he was supposed to use in the tool to fix the check. Something like
> 
> ```lisp
> Update your workflow using the "Restrict permissions for GITHUB_TOKEN" at <link>
> ```
> 
> Or maybe adding the "fix it yourself as a turn around" with something like:
> 
> ```lisp
> Update your workflow using <link> or set the permission default to read (`contents: read`) giving write permissions only to jobs
> ```
> 
> Both options I think would be great.

Can you take a look at the changes I made? You can edit the PR yourself as well by using the branch on my repo, so feel free to make additional changes you think would be meaningful. Let me know once you've made the changes. Thanks!

joycebrum commented on checks/evaluation/permissions/GitHubWorkflowPermissionsTopNoWrite.yml:
```diff
@@ -0,0 +1,20 @@
+    - Tick the 'Restrict permissions for GITHUB_TOKEN'
+    - Untick other options
+  markdown:
+    - Update your workflow using [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/${{ repo }}/${{ workflow }}/${{ branch }}?enable=permissions).
```
I think thats great! The following suggestion is just to keep the markdown equal to the text, besides I think "visit" works better either, considering that there are other two steps.

```suggestion
    - Visit [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/${{ repo }}/${{ workflow }}/${{ branch }}?enable=permissions).
```

laurentsimon commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,16 @@
+
+implementation: >
+  The rule is implemented by checking whether the `permissions` keyword is given non-write permission for the following
+  scopes: X, X, X.
```
@joycebrum do you want to take a stab at this one?

laurentsimon replied:
fyi @olivekl we're revamping the description and remediation

joycebrum replied:
Sure, I'll try to suggest something by tomorrow!

joycebrum commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,16 @@
+short: Checks that GitHub workflows do not have steps with dangerous write permissions
+desc: This rule checks that GitHub workflows do not have steps with dangerous write permissions
+motivation: >
+  TODO: attack surface is the Actions called in the steps, and its dependencies.
```
Not sure I got he TODO right, see what you think 

```suggestion
  Even with permissions default set to read, some scopes having write permissions brings vulnerabilities to the project,
  such as commiting unreviewed code, remove pre-submit checks to introduce a bug, etc.
  For more information about the scopes and the vulnerabilities involved, see https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions.
```

joycebrum commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,16 @@
+  TODO: attack surface is the Actions called in the steps, and its dependencies.
+
+implementation: >
+  The rule is implemented by checking whether the `permissions` keyword is given non-write permission for the following
+  scopes: X, X, X.
+  TODO: permissions for know action to publish are not reported.
```
```suggestion
  The rule is implemented by checking whether the `permissions` keyword is given non-write permission for the following
  scopes: `statuses`, `checks`, `security-events`, `deployments`, `contents`, `packages`, `actions`.
  Write permissions given to recognized packaging actions or commands are allowed and do not bring much risk to the project.
```

joycebrum commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,16 @@
+risk: Medium
+remediation:
+  effort: High
+  text:
+    - Verify which permissions are needed and it you can reduce them. TODO(use annotation)
+  markdown:
+    - Verify which permissions are needed and it you can reduce them. TODO(use annotation)
```
Just a small fix
```suggestion
  text:
    - Verify which permissions are needed and if you can reduce them. TODO(use annotation)
  markdown:
    - Verify which permissions are needed and if you can reduce them. TODO(use annotation)
```

joycebrum reviewed (commented):
Some improvements/fixes to the steps write permission response.

spencerschrock commented on pkg/json.go:
```diff
@@ -83,6 +86,32 @@ type JSONScorecardResultV2 struct {
+// nolint: govet
+type jsonCheckResultV3 struct {
+	Risk     rules.Risk        `json:"risk"`
+	Outcome  finding.Outcome   `json:"outcome"`
```
is your intention to simplify checks to pass/fail? Does a check fail when any rule fails?

laurentsimon replied:
I think that's correct. It's jut a summary of the overall result. Users can parse each rule's result to find out what's passing or not. User can ignore the check's `Outcome` and look at the findings.

Any concern?

spencerschrock commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,18 @@
+motivation: >
+  Even with permissions default set to read, some scopes having write permissions brings vulnerabilities to the project,
+  such as commiting unreviewed code, remove pre-submit checks to introduce a bug, etc.
+  For more information about the scopes and the vulnerabilities involved, see https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions.
```
if we use this as a source of truth to generate checks.md, this may be a circular reference.

laurentsimon replied:
Good point. No plan yet to use these to generate checks.md yet. I'll keep this in mind when when we get to this part (in a later PR)

spencerschrock commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,19 @@
+desc: This rule checks that GitHub workflows do not have steps with dangerous write permissions
+motivation: >
+  Even with permissions default set to read, some scopes having write permissions in their steps brings incurs a risk to the project.
+  By giving write permission to the Actions you call in jobs, external Action you call could abuse them - deendencin on the permissions, this could let them
```
`deendencin` typo

joycebrum replied:
```suggestion
  By giving write permission to the Actions you call in jobs, external Action you call could abuse them - depending on the permissions, this could let them
```

laurentsimon replied:
done, thanks

spencerschrock commented on rule/rule.go:
```diff
@@ -0,0 +1,172 @@
+	return v > vv
+}
+
+func RuleNew(loc embed.FS, rule string) (*Rule, error) {
```
same nit: `New` is less repetitive in my opinion.

`rule.RuleNew` vs `rule.New` when calling from other packages

laurentsimon replied:
done, thanks

spencerschrock reviewed (commented):
Some high level questions that didn't fit elsewhere:

How do rules work across platforms? I assume GitHubWorkflowPermissionsStepsNoWrite will only work on GitHub. Will there be a similar one for Gitlab if the support lands?

By breaking the checks into rules, is there any thought into running subsets of the rules for each check? Whether this is someone who only cares about certain checks, or if you're trying to run local analysis where some rules don't work locally?

laurentsimon commented:
> Some high level questions that didn't fit elsewhere:
> 
> How do rules work across platforms? I assume GitHubWorkflowPermissionsStepsNoWrite will only work on GitHub. Will there be a similar one for Gitlab if the support lands?

Good question. I see two ways:
1. We keep the rule names generic and set the outcome to `Unsupported` if they don't apply
2. We keep the rule names specific to each platform when it's platform-specific (which I think workflow permissions are). Since we want the remediation to be very specific to rules, this allows us to have different remediation text for each provider 

I lean towards option (2) for rules that are CVS-specific.

> By breaking the checks into rules, is there any thought into running subsets of the rules for each check? Whether this is someone who only cares about certain checks, or if you're trying to run local analysis where some rules don't work locally?

we can do that, yes. I think this is beyond this PR, though. To do that, we may need to remove all the plumbing around checks. Right now the entire code revolves around checks. This is a bigger effort. I'm curious though: is local analysis something that is important or not for Scorecard? And how important? What are the main use cases of local analysis for users?

laurentsimon commented on checks/evaluation/permissions/permissions.go:
```diff
@@ -115,8 +129,31 @@ func applyScorePolicy(results *checker.TokenPermissionsData, c *checker.CheckReq
 
-func warnWithRemediation(logger checker.DetailLogger, msg *checker.LogMessage, rem *checker.Remediation) {
-	msg.Remediation = rem
+func createMsg(loct *checker.PermissionLocation) (*checker.LogMessage, error) {
```
updated to `createText()` and `createLogMsg()`

laurentsimon commented on checks/evaluation/permissions/permissions.go:
```diff
@@ -59,30 +65,38 @@ func applyScorePolicy(results *checker.TokenPermissionsData, c *checker.CheckReq
 
-			if msg.Path != "" {
-				rem = remediationMetadata.CreateWorkflowPermissionRemediation(r.File.Path)
+			if r.File.Path != "" {
+				loc.Value = r.File.Path
```
+1. There was a pointer involved at some point and I forgot to update. Thanks!

spencerschrock commented on checks/evaluation/permissions/permissions.go:
```diff
@@ -59,30 +65,38 @@ func applyScorePolicy(results *checker.TokenPermissionsData, c *checker.CheckReq
+			}
+			if r.File.Snippet != "" {
+				loc.Snippet = &r.File.Snippet
+			}
```
I could see it either way. I think I was thinking the raw results could be grouped based on the rules they support. 

But under the idea that a finding is an opinionated result based on the raw data, I'm fine with it as is.

laurentsimon commented:
friendly ping for more reviews? @spencerschrock if you think your comments are addressed, feel free to LGTM. If you think we should continue some discussion, ping the relevant comment.

olivekl commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,33 @@
+  text:
+    - Verify which permissions are needed and if you can reduce them. TODO(use annotation)
+  markdown:
+    - Verify which permissions are needed and if you can reduce them. TODO(use annotation)
```
```suggestion
    - Verify which permissions are needed and consider whether you can reduce them. TODO(use annotation)
```

laurentsimon commented on checks/evaluation/permissions/GitHubWorkflowPermissionsStepsNoWrite.yml:
```diff
@@ -0,0 +1,33 @@
+# limitations under the License.
+
+short: Checks that GitHub workflows do not have steps with dangerous write permissions
+desc: This rule checks that GitHub workflows do not have steps with dangerous write permissions
```
Ideally yes. But I could not come up with a richer description. Maybe it's not needed to have both?

laurentsimon replied:
Will iterate on this over time. This is internal.

spencerschrock commented:
Based on what @varunsh-coder mentioned in today's meeting, I think enhancing the SARIF output could be a good way of exposing these structured results, instead of an `ejson` format.

https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning
> Any valid SARIF 2.1.0 output file can be uploaded, however, code scanning will only use the following supported properties.

Has an [array of `rules` objects](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#toolcomponent-object) which correspond to the array of findings being proposed:
>  Required. An array of reportingDescriptor objects that represent rules. The analysis tool uses rules to find problems in the code being analyzed


## Comparison to current output
Taking a look at the result struct and [commonly used SARIF objects](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317418)
```
type JSONScorecardResultV3 struct {
	Date           string              `json:"date"`
	Repo           jsonRepoV2          `json:"repo"`
	Scorecard      jsonScorecardV2     `json:"scorecard"`
	AggregateScore jsonFloatScore      `json:"score"`
	Checks         []jsonCheckResultV3 `json:"checks"`
	Metadata       []string            `json:"metadata"`
}
```

`Date`: [startTimeUtc property](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317574)
> An invocation object MAY contain a property named startTimeUtc whose value is a string in the format specified in §3.9, specifying the UTC date and time at which the invocation started.

`Repo`: Might need to be included in a [property bag](https://github.com/microsoft/sarif-tutorials/blob/main/docs/2-Basics.md#property-bags)

`Scorecard`:  The [toolComponentObject](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317533) seems like it has [SEMVER](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317545) and [Version](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317546) fields which would cover the version and commit fields.

`AggregateScore`: again, maybe [property bag](https://github.com/microsoft/sarif-tutorials/blob/main/docs/2-Basics.md#property-bags)

`Checks`: A lot of similarities if you compare with a [minimal example](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#example-with-minimum-required-properties). Grouping by check may require some thought, maybe a run object for each check?

`metadata`: again, maybe [property bag](https://github.com/microsoft/sarif-tutorials/blob/main/docs/2-Basics.md#property-bags)


I personally like this approach as:
* It's a standard some static analysis tools are converging on
* we already support SARIF with the Scorecard Action so improving it would improve our code scanning integration for "free".

laurentsimon commented:
@varunsh-coder you also mentioned codeQL rule naming convention, containing the risk. Could you paste the link?

In the sync, folks were interested in using numbers rather than strings for fields like `Risk` and `Outcome`.

laurentsimon commented:
> Checks: A lot of similarities if you compare with a [minimal example](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#example-with-minimum-required-properties). Grouping by check may require some thought, maybe a run object for each check?

Would we be able to represent SC rules? IIUC, SARIF rules cannot have nested rules, so maybe it would be `SC rule = SARIF rule` mapping and a property-bad field to indicate the check it belongs to?

Some question about SARIF:
1. Can we express positive results? For example, our results today do report positive results. If we want to keep them in our result, can we report them with SARIF? I think SARIF was designed to report problems, so this may require using a `severity = none` and not break existing tooling
2. SARIF only understand path of type "files" inside the repo. You cannot link to anything outside, like we support using our URL type for the location.
3. GitHub SARIF always requires a location. This is awkward because results like branch protection, etc don't have a file associated with it. To make it work on GitHub, we provide a path to our config file. This is just a hack
4. Are remediation fields supported in SARIF? I don't see one, but maybe? For GitHub, we ended up concatenating the text and the remediation text. In this PR, we want to support not only remediation, but also automated `Patch` field to allow downstream tools to create PRs automatically.

I'm still not convinced about SARIF. What I like about it is that they tried to structure the Path / Location. What I don't like is that it's complicated to write, to read. The specs is big, no tool supports it well, it does not fit well with scorecard (alerts with GH API, positive results, remediation) unless we use property bags everywhere. In the end, I feel we don't get much out of it, except for a source of inspiration.

> we already support SARIF with the Scorecard Action so improving it would improve our code scanning integration for "free".

I'm not sure we get anything for free. No tool would understand our format, neither would GH: they backend verification might fail if we don't respect all the constraints they've added for *their* use case. This adds a lot of constraints which we can avoid by defining a simple JSON format of our choice. In a nutshell, SARIF is *a* standard but you need to comply with the consumer's expectations / constraints to make it work, and in practice you end up custom SARIF implementation for multiple consumers. 

Let's keep this conversation going.

spencerschrock commented:
> > Checks: A lot of similarities if you compare with a [minimal example](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#example-with-minimum-required-properties). Grouping by check may require some thought, maybe a run object for each check?
> 
> Would we be able to represent SC rules? IIUC, SARIF rules cannot have nested rules, so maybe it would be `SC rule = SARIF rule` mapping and a property-bad field to indicate the check it belongs to?

My thought was a run object for each check since SARIF files are an array of runs, but there's also property bags and tag fields.

> Some question about SARIF:
> 
> 1. Can we express positive results? For example, our results today do report positive results. If we want to keep them in our result, can we report them with SARIF? I think SARIF was designed to report problems, so this may require using a `severity = none` and not break existing tooling

Yes. https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317647

> 2. SARIF only understand path of type "files" inside the repo. You cannot link to anything outside, like we support using our URL type for the location.

You're talking about CI/CD runs for the SAST check and things like that?

> 3. GitHub SARIF always requires a location. This is awkward because results like branch protection, etc don't have a file associated with it. To make it work on GitHub, we provide a path to our config file. This is just a hack

Wasn't aware of this.

> 4. Are remediation fields supported in SARIF? I don't see one, but maybe? For GitHub, we ended up concatenating the text and the remediation text. Although not implemented in this PR, we want to ultimately support not only remediation text, but also automated `Patch` field to allow downstream tools to create PRs automatically.

Yes https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317881

> I'm still not convinced about SARIF. What I like about it is that they tried to structure the Path / Location. What I don't like is that it's complicated to generate and to parse (they use indexes for referring to rules, for examples). The specs are "big", no tool supports it well, it does not fit well with scorecard (alerts with GH API, positive results, remediation) unless we use property bags everywhere. In the end, I feel we don't get much out of it, except for a source of inspiration.

> I'm not sure we get anything for free. No tool would understand our format, neither would GH: they backend verification might fail if we don't respect all the constraints they've added for _their_ use case. This adds a lot of constraints which we can avoid by defining a simple JSON format of our choice. In a nutshell, SARIF is _a_ standard but you need to comply with the consumer's expectations / constraints to make it work, and in practice you may end up with custom SARIF implementation for each consumer.

At the end of the day, I see structured results making it easier to improve the SARIF we already have. I'm not advocating that the results have to exclusively be in SARIF, but at the very least our SARIF results could get better after we have structured results. But that might be outside the scope of this PR.

laurentsimon commented:
Thanks @spencerschrock for all the pointers and answers !

> > Would we be able to represent SC rules? IIUC, SARIF rules cannot have nested rules, so maybe it would be `SC rule = SARIF rule` mapping and a property-bad field to indicate the check it belongs to?
> 
> My thought was a run object for each check since SARIF files are an array of runs, but there's also property bags and tag fields.

I think we could improve the existing SARIF format using the SC rules, as it would allow more granular alerts. For example, instead of reporting a branch protection alert, we can have one alert per setting, and users could close / ignore each of them independently. With our existing implementation: if a user dismisses the single BP alert (because they remediated all the settings they care about) and later changes a BP setting, no alert will ever be created.  It'd be nice to improve this behavior.

Wdut?

> 
> > Some question about SARIF:
> > 
> > 1. Can we express positive results? For example, our results today do report positive results. If we want to keep them in our result, can we report them with SARIF? I think SARIF was designed to report problems, so this may require using a `severity = none` and not break existing tooling
> 
> Yes. https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317647

+1, that's great! Probably need to verify if GH supports this, but that's good to know.

> > 2. SARIF only understand path of type "files" inside the repo. You cannot link to anything outside, like we support using our URL type for the location.
> 
> You're talking about CI/CD runs for the SAST check and things like that?

Yes. Or, for example, when a file is outside the repository under analysis, e.g. the security policy defined by the org. This particular example is not a problem because GH dashboard need not show positive results, though. ANother example could be if we don't find fuzzing on ossf-fuzz repo: we may want to link to the github.com/google/oss-fuzz/path/to/fuzz-targets as the "location".

> 
> > 3. GitHub SARIF always requires a location. This is awkward because results like branch protection, etc don't have a file associated with it. To make it work on GitHub, we provide a path to our config file. This is just a hack
> 
> Wasn't aware of this.
> 
> > 4. Are remediation fields supported in SARIF? I don't see one, but maybe? For GitHub, we ended up concatenating the text and the remediation text. Although not implemented in this PR, we want to ultimately support not only remediation text, but also automated `Patch` field to allow downstream tools to create PRs automatically.
> 
> Yes https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317881

Good to know, thanks. Does this support `git diff` to apply fixes (https://linux.die.net/man/1/patch)? That's how I'd imagine the automated remediation via PR to work. 

> At the end of the day, I see structured results making it easier to improve the SARIF we already have. 

I would actually like to use the SC rule as SARIF rule IDs to make the GH SARIF alerts more granular. So I'm definitely onboard for improving the GH SARIF results at some point.

> I'm not advocating that the results have to exclusively be in SARIF, but at the very least our SARIF results could get better after we have structured results. But that might be outside the scope of this PR.

Makes sense. I thought @varunsh-coder was asking whether we should use SARIF as our only format

varunsh-coder commented on checks/evaluation/permissions/GitHubWorkflowPermissionsTopNoWrite.yml:
```diff
@@ -0,0 +1,34 @@
+    - Tick the 'Restrict permissions for GITHUB_TOKEN'
+    - Untick other options
+  markdown:
+    - Visit [https://app.stepsecurity.io/secureworkflow](https://app.stepsecurity.io/secureworkflow/${{ repo }}/${{ workflow }}/${{ branch }}?enable=permissions).
```
I want to suggest adding in the remediation text that to remediate multiple issues across different files in the repo, visit https://app.stepsecurity.io/securerepo. That way, various issues (token permissions, pinning dependencies, Dependabot config, adding CodeQL) can be fixed in a single pull request. The PR is created using a fork, and no App is needed to be installed. Many users already use this option but have to discover it themselves. Examples of PRs can be found [here](https://github.com/step-security/secure-workflows).

laurentsimon replied:
Thanks for the suggestion. I'll try to add this into the text

naveensrinivasan commented on rule/rule.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	if err := validate(r); err != nil {
+		return nil, err
```
Nit, return `fmt.Errorf`

naveensrinivasan replied:
This is applicable in other files.

laurentsimon replied:
`validate()` does use `fmt.Errorf` already.

naveensrinivasan commented on pkg/common.go:
```diff
@@ -33,20 +33,45 @@ func DetailToString(d *checker.CheckDetail, logLevel log.Level) string {
-	sb.WriteString(fmt.Sprintf("%s: %s", typeToString(d.Type), d.Msg.Text))
+	// Non-structured results.
+	//nolint: nestif
+	if d.Msg.Finding == nil {
```
This is hard to maintain with this many nesting. Can this be broken into separate funcs?

laurentsimon replied:
done, thanks

naveensrinivasan reviewed (approved):
This is cool! Thanks. A few nits.

laurentsimon commented:
I addressed all the comments. Any last comments?

NOTE: this PR just creates the plumbing to start migrating other checks. We will iterate over the rule definition for the Token-Permissions in follow-up PRs. We may also add a list of RuleID runs for the check.

laurentsimon commented on finding/finding.go:
```diff
@@ -0,0 +1,122 @@
+	// Outcome is positive, no remediation needed.
+	f.Remediation = nil
+	return f
+}
```
done. We can change this in the future. I was trying to avoid copying complex structure with pointers.

laurentsimon commented on finding/finding.go:
```diff
@@ -0,0 +1,122 @@
+	return f
+}
+
+func (f *Finding) WithRemediationMetadata(values map[string]string) *Finding {
```
good point. I added a warning int he function description.

laurentsimon commented on finding/finding_test.go:
```diff
@@ -0,0 +1,204 @@
+	patch := "some patch values"
+	sline := uint(10)
+	eline := uint(46)
+	positiveOutcome := OutcomePositive
```
I think so, because I cannot use `&OutcomePositive` with global values :/
If you know how to do it, I'll use it.

laurentsimon commented on finding/finding_test.go:
```diff
@@ -0,0 +1,204 @@
+		{
+			name: "location",
+			id:   "testdata/env-variables",
+			env:  map[string]string{"branch": "master", "repo": "ossf/scorecard"},
```
renamed ot metadata

laurentsimon commented on checker/raw_result.go:
```diff
@@ -286,11 +287,11 @@ type ArchivedStatus struct {
-	Snippet   string   // Snippet of code
-	Offset    uint     // Offset in the file of Path (line for source/text files).
-	EndOffset uint     // End of offset in the file, e.g. if the command spans multiple lines.
-	FileSize  uint     // Total size of file.
-	Type      FileType // Type of file.
+	Snippet   string           // Snippet of code
+	Offset    uint             // Offset in the file of Path (line for source/text files).
+	EndOffset uint             // End of offset in the file, e.g. if the command spans multiple lines.
+	FileSize  uint             // Total size of file.
+	Type      finding.FileType // Type of file.
 	// TODO: add hash.
 }
```
agreed. Since the raw results are not public yet, we could merge the two structure and use the same. Filed https://github.com/ossf/scorecard/issues/2626

laurentsimon commented on finding/finding.go:
```diff
@@ -0,0 +1,122 @@
+
+// Location represents the location of a finding.
+// nolint: govet
+type Location struct {
+	Type      FileType `json:"type"`
+	Value     string   `json:"value"`
+	LineStart *uint    `json:"lineStart,omitempty"`
+	LineEnd   *uint    `json:"lineEnd,omitempty"`
+	Snippet   *string  `json:"snippet,omitempty"`
+}
```
filed https://github.com/ossf/scorecard/issues/2626
